File: internal/pricing/pricing_test.go
package pricing

import "testing"

func TestExactMatch(t *testing.T) {
	r, ok := Lookup("gpt-4o")
	if !ok {
		t.Fatal("exact match should work")
	}
	if r.Input != 2.5 {
		t.Fatalf("gpt-4o input rate = %v, want 2.5", r.Input)
	}
}

func TestPrefixMatchDeterministic(t *testing.T) {
	// gpt-4o-mini should ALWAYS match gpt-4o-mini (input=0.15), never
	// gpt-4o (input=2.5). This was non-deterministic before the fix.
	for i := 0; i < 100; i++ {
		r, ok := Lookup("gpt-4o-mini-2024")
		if !ok {
			t.Fatal("prefix match should work")
		}
		if r.Input != 0.15 {
			t.Fatalf("iteration %d: gpt-4o-mini matched wrong rate (input=%v, want 0.15)", i, r.Input)
		}
	}

	for i := 0; i < 100; i++ {
		r, ok := Lookup("o3-mini-2024")
		if !ok {
			t.Fatal("prefix match should work")
		}
		if r.Input != 1.1 {
			t.Fatalf("iteration %d: o3-mini matched wrong rate (input=%v, want 1.1)", i, r.Input)
		}
	}
}

func TestUnknownModel(t *testing.T) {
	_, ok := Lookup("nonexistent-model-xyz")
	if ok {
		t.Fatal("unknown model should return false")
	}
}

func TestCostCalculation(t *testing.T) {
	// 1M input tokens of gpt-4o = $2.5
	c := Cost("gpt-4o", 1_000_000, 0, 0, 0)
	if c != 2.5 {
		t.Fatalf("1M input of gpt-4o = %.4f, want 2.5", c)
	}

	// 1M input + 1M output of claude-sonnet-4-5 = $3 + $15 = $18
	c = Cost("claude-sonnet-4-5", 1_000_000, 1_000_000, 0, 0)
	if c != 18 {
		t.Fatalf("1M in + 1M out of claude-sonnet-4-5 = %.4f, want 18", c)
	}
}

File: tests/pricing.test.ts
import { describe, expect, test } from "bun:test";
import { lookup, cost } from "../src/pricing/pricing.js";

describe("pricing", () => {
  test("exact match", () => {
    const r = lookup("claude-sonnet-4-5");
    expect(r).not.toBeNull();
    expect(r!.input).toBe(3);
    expect(r!.output).toBe(15);
  });

  test("prefix match: claude-sonnet-4-5-20251020 matches claude-sonnet-4-5", () => {
    const r = lookup("claude-sonnet-4-5-20251020");
    expect(r).not.toBeNull();
    expect(r!.input).toBe(3);
  });

  test("longest prefix wins: gpt-4o-mini matches gpt-4o-mini, not gpt-4o", () => {
    const r = lookup("gpt-4o-mini");
    expect(r).not.toBeNull();
    expect(r!.input).toBe(0.15);
  });

  test("opencode-go models are present", () => {
    expect(lookup("glm-5.2")).not.toBeNull();
    expect(lookup("deepseek-v4-pro")).not.toBeNull();
    expect(lookup("kimi-k2.7-code")).not.toBeNull();
    expect(lookup("qwen3.7-max")).not.toBeNull();
    expect(lookup("minimax-m3")).not.toBeNull();
    expect(lookup("mimo-v2.5-pro")).not.toBeNull();
  });

  test("unknown model returns null", () => {
    expect(lookup("unknown-model-xyz")).toBeNull();
  });

  test("cost calculation", () => {
    // 1M input tokens of claude-sonnet-4-5 = $3
    expect(cost("claude-sonnet-4-5", 1_000_000, 0)).toBeCloseTo(3, 6);
    // 1M output tokens = $15
    expect(cost("claude-sonnet-4-5", 0, 1_000_000)).toBeCloseTo(15, 6);
    // Unknown model = 0
    expect(cost("unknown", 1000, 1000)).toBe(0);
  });

  test("cost includes cache read/write", () => {
    // claude-sonnet-4-5: cache_read=$0.30/M, cache_write=$3.75/M
    const c = cost("claude-sonnet-4-5", 0, 0, 1_000_000, 1_000_000);
    expect(c).toBeCloseTo(0.3 + 3.75, 6);
  });
});
